### PR TITLE
Upload unified

### LINKED ISSUE #188: Allow https Mirrors
Might already been fixed with the new Upload System. However in the latest release the only protocols allowed for mirror links are `ftp` and `http`.

### PR DESCRIPTION
**Pull request by Phhere**
New Upload Class which checks mimetypes and reports correct error messages

fixes #188
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/webSPELL/webSPELL/pull/190%23issuecomment-82668198%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23issuecomment-89267569%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23issuecomment-92433938%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23issuecomment-93008675%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r23914501%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24021340%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24035233%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24035498%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24036096%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24037558%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24039577%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24300053%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24300054%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24300056%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24300061%22%2C%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r27726206%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23issuecomment-82668198%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22https%3A//github.com/webSPELL/webSPELL/blob/upload_unified/files.php%23L1153%5Cr%5Cn%5Cr%5CnThere%20is%20a%20%24i%20which%20is%20not%20defined.%5Cr%5CnNotice%3A%20Undefined%20variable%3A%20i%20in%20/var/www/webspell.test/htdocs/files.php%20on%20line%201153%20%22%2C%20%22created_at%22%3A%20%222015-03-18T01%3A30%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/706758%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/derchrisuk%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-04-03T12%3A03%3A42Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415280%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Phhere%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3Ashipit%3A%22%2C%20%22created_at%22%3A%20%222015-04-13T17%3A21%3A42Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415280%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Phhere%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3Ashipit%3A%22%2C%20%22created_at%22%3A%20%222015-04-14T18%3A27%3A06Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/706758%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/derchrisuk%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%2046f5956864bbb68b4d8940cc840b0e1be786e510%20admin/bannerrotation.php%20478%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24036096%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Allow%20https%22%2C%20%22created_at%22%3A%20%222015-02-03T20%3A21%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/498197%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Pascalmh%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-02-08T17%3A05%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415280%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Phhere%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20admin/bannerrotation.php%3AL368-416%22%7D%2C%20%22Pull%2046f5956864bbb68b4d8940cc840b0e1be786e510%20demos.php%20654%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24039577%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Why%20this%20change%3F%22%2C%20%22created_at%22%3A%20%222015-02-03T21%3A04%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/498197%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Pascalmh%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222015-02-08T17%3A05%3A47Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415280%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Phhere%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20demos.php%3AL657-665%22%7D%2C%20%22Pull%2046f5956864bbb68b4d8940cc840b0e1be786e510%20admin/bannerrotation.php%205%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24035233%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Remove%20whitespace%20as%20it%20might%20hunt%20us%20in%20the%20future%20%28Search%20/%20Replace%29%22%2C%20%22created_at%22%3A%20%222015-02-03T20%3A11%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/498197%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Pascalmh%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-none%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-02-08T17%3A05%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415280%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Phhere%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20admin/bannerrotation.php%3AL23-48%22%7D%2C%20%22Pull%20460dfaee6d015dafd3ef12e73cffa6f55de76f99%20files.php%2035%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r23914501%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Can%20we%20fix%20this%20here%20as%20well%20%23188%3F%22%2C%20%22created_at%22%3A%20%222015-02-02T09%3A44%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/498197%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Pascalmh%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%22%2C%20%22created_at%22%3A%20%222015-02-03T17%3A24%3A40Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/498197%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Pascalmh%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20files.php%3AL97-318%22%7D%2C%20%22Pull%2046f5956864bbb68b4d8940cc840b0e1be786e510%20admin/bannerrotation.php%20164%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24035498%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Allow%20https%20%28or%20just%20use%20%60//%60%3F%29%22%2C%20%22created_at%22%3A%20%222015-02-03T20%3A14%3A49Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/498197%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Pascalmh%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-complete%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-02-08T17%3A05%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415280%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Phhere%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20admin/bannerrotation.php%3AL101-158%22%7D%2C%20%22Pull%2046f5956864bbb68b4d8940cc840b0e1be786e510%20admin/sponsors.php%20430%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/webSPELL/webSPELL/pull/190%23discussion_r24037558%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22DRY%22%2C%20%22created_at%22%3A%20%222015-02-03T20%3A39%3A43Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/498197%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Pascalmh%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3A%2B1%3A%3Ca%20href%3D%27%23crh-update-complete%27%3E%3C/a%3E%22%2C%20%22created_at%22%3A%20%222015-04-03T12%3A03%3A43Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415280%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/Phhere%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20admin/sponsors.php%3AL257-372%22%7D%7D%2C%20%22approved%22%3A%20%7B%22https%3A//github.com/derchrisuk%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/706758%3Fv%3D3%22%7D%2C%20%22https%3A//github.com/Phhere%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415280%3Fv%3D3%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>

<a href='https://github.com/derchrisuk'></a><a href='https://github.com/Phhere'></a>

- [x] <a href='#crh-comment-Pull 460dfaee6d015dafd3ef12e73cffa6f55de76f99 files.php 35'></a> &nbsp;<b><a href='https://github.com/webSPELL/webSPELL/pull/190#discussion_r23914501'>File: files.php:L97-318</a></b>
- <a href='https://github.com/Pascalmh'></a> Can we fix this here as well #188?
- [x] <a href='#crh-comment-Pull 46f5956864bbb68b4d8940cc840b0e1be786e510 admin/bannerrotation.php 5'></a> &nbsp;<b><a href='https://github.com/webSPELL/webSPELL/pull/190#discussion_r24035233'>File: admin/bannerrotation.php:L23-48</a></b>
- <a href='https://github.com/Pascalmh'></a> Remove whitespace as it might hunt us in the future (Search / Replace)
- [x] <a href='#crh-comment-Pull 46f5956864bbb68b4d8940cc840b0e1be786e510 admin/bannerrotation.php 164'></a> &nbsp;<b><a href='https://github.com/webSPELL/webSPELL/pull/190#discussion_r24035498'>File: admin/bannerrotation.php:L101-158</a></b>
- <a href='https://github.com/Pascalmh'></a> Allow https (or just use `//`?)
- [x] <a href='#crh-comment-Pull 46f5956864bbb68b4d8940cc840b0e1be786e510 admin/bannerrotation.php 478'></a> &nbsp;<b><a href='https://github.com/webSPELL/webSPELL/pull/190#discussion_r24036096'>File: admin/bannerrotation.php:L368-416</a></b>
- <a href='https://github.com/Pascalmh'></a> Allow https
- [x] <a href='#crh-comment-Pull 46f5956864bbb68b4d8940cc840b0e1be786e510 admin/sponsors.php 430'></a> &nbsp;<b><a href='https://github.com/webSPELL/webSPELL/pull/190#discussion_r24037558'>File: admin/sponsors.php:L257-372</a></b>
- <a href='https://github.com/Pascalmh'></a> DRY
- [x] <a href='#crh-comment-Pull 46f5956864bbb68b4d8940cc840b0e1be786e510 demos.php 654'></a> &nbsp;<b><a href='https://github.com/webSPELL/webSPELL/pull/190#discussion_r24039577'>File: demos.php:L657-665</a></b>
- <a href='https://github.com/Pascalmh'></a> Why this change?
- [x] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/webSPELL/webSPELL/pull/190#issuecomment-82668198'>General Comment</a></b>
- <a href='https://github.com/derchrisuk'></a> https://github.com/webSPELL/webSPELL/blob/upload_unified/files.php#L1153
There is a $i which is not defined.
Notice: Undefined variable: i in /var/www/webspell.test/htdocs/files.php on line 1153

<a href='https://www.codereviewhub.com/webSPELL/webSPELL/pull/190?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/webSPELL/webSPELL/pull/190?approve=1'></a>&nbsp;<a href='https://www.codereviewhub.com/webSPELL/webSPELL/pull/190?approve=0'></a>&nbsp;<a href='https://github.com/webSPELL/webSPELL/pull/190'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: files.php [hunk 1]
```diff
@@ -97,206 +97,221 @@ function unit_to_size($num, $unit)
...
+        $filename = $_POST[ 'filename' ];
+        $fileurl = $_POST[ 'fileurl' ];
+        $filesize = unit_to_size($_POST[ 'filesize' ], $_POST[ 'unit' ]);
+        $info = $_POST[ 'info' ];
+        $accesslevel = $_POST[ 'accesslevel' ];
+        $mirror1 = $_POST[ 'mirror2' ];
+        $mirror2 = $_POST[ 'mirror3' ];
+
+        // MIRRORS
+        if (stristr($mirror1, "http://") || stristr($mirror1, "ftp://")) {
```
Comment (Pascalmh) [inline files.php]: Can we fix this here as well #188?

Comment (Pascalmh) [inline files.php]: :+1:

File: admin/bannerrotation.php [hunk 2]
```diff
@@ -23,25 +23,25 @@
 #   visit webspell.org                                                   #
 #                                                                        #
 ##########################################################################
-*/
+ */
```
Comment (Pascalmh) [inline admin/bannerrotation.php]: Remove whitespace as it might hunt us in the future (Search / Replace)

File: admin/bannerrotation.php [hunk 3]
```diff
@@ -105,59 +101,57 @@
...
 
+    $upload = new \webspell\Upload('banner');
+
     $CAPCLASS = new \webspell\Captcha;
-    if ($CAPCLASS->checkCaptcha(0, $_POST[ 'captcha_hash' ])) {
-        if ($bannername && $bannerurl && $banner) {
-            if (!stristr($bannerurl, 'http://')) {
+    if ($CAPCLASS->checkCaptcha(0, $_POST['captcha_hash'])) {
+        if ($bannername && $bannerurl) {
+            if (stristr($bannerurl, 'http://')) {
```
Comment (Pascalmh) [inline admin/bannerrotation.php]: Allow https

File: admin/bannerrotation.php [hunk 4]
```diff
@@ -335,46 +368,48 @@
...
-                $displayed = '<font color="green"><b>' . $_language->module[ 'yes' ] . '</b></font>' :
-                $displayed = '<font color="red"><b>' . $_language->module[ 'no' ] . '</b></font>';
+            if ($ds['displayed'] == 1) {
+                $displayed = '<font color="green"><b>' . $_language->module['yes'] . '</b></font>';
+            } else {
+                $displayed = '<font color="red"><b>' . $_language->module['no'] . '</b></font>';
+            }
 
-            if (stristr($ds[ 'bannerurl' ], 'http://')) {
+            if (stristr($ds['bannerurl'], 'http://')) {
```
Comment (Pascalmh) [inline admin/bannerrotation.php]: Allow https

File: admin/sponsors.php [hunk 5]
```diff
@@ -264,132 +257,115 @@ function chkFormular() {
...
+        $upload = new \webspell\Upload('banner_small');
+        if ($upload->hasFile()) {
+            if ($upload->hasError() === false) {
+                $mime_types = array('image/jpeg','image/png','image/gif');
+
+                if ($upload->supportedMimeType($mime_types)) {
+                    $imageInformation =  getimagesize($upload->getTempFile());
+
+                    if (is_array($imageInformation)) {
+                        switch ($imageInformation[ 2 ]) {
```
Comment (Pascalmh) [inline admin/sponsors.php]: DRY

File: demos.php [hunk 6]
```diff
@@ -675,11 +657,8 @@ function top5()
     }
     $sort = "date";
     if (isset($_GET[ 'sort' ])) {
-        if ($_GET[ 'sort' ] == 'date'
-            || $_GET[ 'sort' ] == 'game'
-            || $_GET[ 'sort' ] == 'league'
-            || $_GET[ 'sort' ] == 'rating'
-            || $_GET[ 'sort' ] == 'downloads'
+        if ($_GET[ 'sort' ] == 'date' || $_GET[ 'sort' ] == 'game' || $_GET[ 'sort' ] == 'league' ||
```
Comment (Pascalmh) [inline demos.php]: Why this change?

File: admin/bannerrotation.php (hunk 4 above)
Comment (Phhere) [inline admin/bannerrotation.php]: :+1:<a href='#crh-update-none'></a>

File: admin/bannerrotation.php (hunk 2 above)
Comment (Phhere) [inline admin/bannerrotation.php]: :+1:<a href='#crh-update-none'></a>

File: admin/bannerrotation.php (hunk 3 above)
Comment (Phhere) [inline admin/bannerrotation.php]: :+1:<a href='#crh-update-complete'></a>

File: demos.php (hunk 6 above)
Comment (Phhere) [inline demos.php]: :+1:

**Comment by derchrisuk** [comment]
https://github.com/webSPELL/webSPELL/blob/upload_unified/files.php#L1153

There is a $i which is not defined.
Notice: Undefined variable: i in /var/www/webspell.test/htdocs/files.php on line 1153

**Comment by Phhere** [comment]
:+1:<a href='#crh-update-none'></a>

File: admin/sponsors.php (hunk 5 above)
Comment (Phhere) [inline admin/sponsors.php]: :+1:<a href='#crh-update-complete'></a>

**Comment by Phhere** [comment]
:shipit:

**Comment by derchrisuk** [comment]
:shipit:
